# SwarnSeq
This R package performs differential expression and differential zero inflation analysis of single-cell RNA-seq (scRNA-seq) UMI counts data through adjusting cell capture efficiency.

Differential Expression (DE) analysis serves as a key downstream analysis for scRNA-seq data. It is necessary for generating informative gene sets, which can be subsequently utilized as input for various other analyses, including pathway or gene set analysis, gene regulation modeling, and gene network analysis. Additionally, numerous methods and tools exist in the literature for conducting DE analysis of scRNA-seq data, which vary significantly based on their statistical approaches, input data, and other factors. Many of the current methods often overlook cell-level auxiliary information such as cell clusters, cell cycles, and molecular capturing procedures and are restricted to comparing only two cellular populations in their model development.

For example, DEsingle, a widely used method, fails to account for factors such as the molecular capturing process, cellular process, cellular clusters, cell cycle stages, and cell phages in differential expression (DE) analysis and is restricted to comparisons between only two cell groups. Consequently, the SwarnSeq approach and the R package are designed to conduct various analyses, including DE analysis, differential zero inflation analysis, and gene classification into subtypes, utilizing the RNA capturing process and cell-level auxiliary data (i.e., cell clusters and other cell-level information) in the development of statistical models for scRNA-seq data.
