---
title: "SwarnSeq"
author:
  - name: Samarendra Das
    affiliation: ICAR-National Institute on Foot and Mouth Disease
    email: samarendra.das@icar.org.in
  - name: Satyajit Chhatoi
    affiliation: Odisha University of Agriculture & Technology
    email: satyajitchhatoi016@gmail.com
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SwarnSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Session Information

```{r sessionInfo}
utils::sessionInfo()
```

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r, Load_Package, include = FALSE}
# Load the package
library(SwarnSeq)
```

# Introduction

This R package performs differential expression and differential zero inflation analysis of single-cell RNA-seq (scRNA-seq) UMI counts data through adjusting cell capture efficiency.

Differential Expression (DE) analysis serves as a key downstream analysis for scRNA-seq data. It is necessary for generating informative gene sets, which can be subsequently utilized as input for various other analyses, including pathway or gene set analysis, gene regulation modeling, and gene network analysis. Additionally, numerous methods and tools exist in the literature for conducting DE analysis of scRNA-seq data, which vary significantly based on their statistical approaches, input data, and other factors. Many of the current methods often overlook cell-level auxiliary information such as cell clusters, cell cycles, and molecular capturing procedures and are restricted to comparing only two cellular populations in their model development.

For example, DEsingle, a widely used method, fails to account for factors such as the molecular capturing process, cellular process, cellular clusters, cell cycle stages, and cell phages in differential expression (DE) analysis and is restricted to comparisons between only two cell groups. Consequently, the SwarnSeq approach and the R package are designed to conduct various analyses, including DE analysis, differential zero inflation analysis, and gene classification into subtypes, utilizing the RNA capturing process and cell-level auxiliary data (i.e., cell clusters and other cell-level information) in the development of statistical models for scRNA-seq data.

```{r, Install_SwarnSeq, eval = FALSE}
# Install the package from Github.
devtools::install_github("nifmd-bbf/SwarnSeq", build_vignettes = TRUE)
library(SwarnSeq)
```

#### We recommend using ***SwarnSeq::***<Function> to access the functions of this package, instead of using ***library(SwarnSeq)***.

# Information about the functions

Please visit the documentation page to know more about the functionality of each function in detail.

# Some Details about the functions

There are actually two major functions to get the result and these two functions depend on some other functions for internal processing. Those functions are 'SwarnUnadjLRT' and 'SwarnAdjLRT' depend on dzinb and ZINBEM functions. Its recommended that users avoid these dependent functions to save time and get the desired results by running the major functions.

Here's a simple and quick example of these two major functions.
To keep the runtime managable for this demonstration, we'have utilized a sample from the example dataset instead of the full version.

```{r}
library(SwarnSeq)
# Load the test data.
data(TestData)

CountData <- as.matrix(TestData$CountData[1:100,1:50])
CountData <- cbind(CountData, TestData$CountData[1:100,300:349])
group <- c(rep(1,50), rep(2,50))
CellCluster <- c(rep(1,30), rep(2,20), rep(3, 10), rep(4, 15), rep(5, 25))
group <- as.factor(group)
CellCluster <- as.factor(CellCluster)
CellAuxil <- as.factor(c(rep("A",14), rep("B",26), rep("C", 40), rep("E", 20)))  # Optional

SwarnUnadjLRT_Results <- suppressWarnings(SwarnSeq::SwarnUnadjLRT(CountData = CountData, norm.method = "DEseq.norm", group = group, CellCluster = CellCluster, CellAuxil = CellAuxil))

SwarnAdjLRT_Results <- suppressWarnings(SwarnSeq::SwarnAdjLRT(CountData = CountData, norm.method = "DEseq.norm", group = group, CellCluster = CellCluster, CellAuxil = CellAuxil))
```

**SwarnUnadjLRT:** This function performs differential gene expression (DEG) analysis on a single-cell RNA-sequencing (scRNA-seq) raw read counts matrix data without capture efficiency correction.

**SwarnAdjLRT:** This function performs differential gene expression (DEG) analysis on a single-cell RNA-sequencing (scRNA-seq) raw read counts matrix.

# Further Processing of Obtained Results

After obtaining the results, a user can further use the functions SwarnClassDE and SwarnTopTags.

Here's a simple and quick example of these two functions.

```{r}
SwarnClass <- SwarnSeq::SwarnClassDE(results = SwarnAdjLRT_Results, alpha = 0.01)

top_genes <- SwarnSeq::SwarnTopTags(results = SwarnAdjLRT_Results, m = 50)
```

**SwarnClassDE:** This function is used to clasify the influential genes of single-cell RNA-seq (scRNA-seq) data.

**SwarnTopTags**: This function selects the most statistically significant genes in your scRNA-seq data.

# Other Functions

The remaining functions, named 'CapEff' and 'ExtAdjNormData,' are used for the following purposes.

**CapEff:** This function estimates the capture efficiencies of cells from single-cell RNA-seq studies. This function takes ERCC spike-in transcript and molecular concentration data, if available.

**ExtAdjNormData:** This function returns a list of two distinct capture efficiency adjusted normalized counts matrices, based on the number of factors in the group vector for the scRNA-seq data.

```{r}
library(SwarnSeq)
# Load the test data.
data(TestData)

CountData <- as.matrix(TestData$CountData[1:100,1:50])
CountData <- cbind(CountData, TestData$CountData[1:100,300:349])

capeff <- SwarnSeq::CapEff(CountData = CountData, CE.range = c(0.01, 0.05), RNAspike.use = FALSE, method = "")

ExtractedAdjNormData <- SwarnSeq::ExtAdjNormData(CountData = CountData, norm.method = "DEseq.norm", group = group, RNAspike.use = FALSE, CE.range = c(0.01, 0.5))
```

We hope this package helps you get the best Differential Expression Analyses results for your scRNA-seq data and assist you in making the best decisions based on those results.

Thank you.
